### PR TITLE
add train/testing files for xview

### PR DESCRIPTION
**Pull request by Honaker**
PR that adds the training and testing files I used to create the new xview subset model that is used in the xview example file. These were added to the xview folder in the example/object_detection folder.

### CODE REVIEW

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+BATCH_SIZE = 8
+import torch
+
+torch.set_float32_matmul_precision("high")
+#import armory.data.datasets
+import datasets
+from datasets import load_from_disk
+from datasets.filesystems import S3FileSystem
+
+LABEL_MAP = {
```
Comment (treubig26) [inline examples/src/charmory_examples/object_detection/xview/train_xview_model.py:18]: Can we add a comment documenting what we're mapping from and to? And why we have to remap the labels.

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    86: 58,  # Storage Tank
+    89: 59,  # Shipping Container Lot
+    91: 60,  # Shipping Container
+    93: 61,  # Pylon
+    94: 62,  # Tower
+}
+
+def load_huggingface_dataset():
+    s3 = S3FileSystem(anon=False)
+    train_dataset = load_from_disk("s3://armory-library-data/datasets/train/", fs=s3)
```
Comment (treubig26) [inline examples/src/charmory_examples/object_detection/xview/train_xview_model.py:85]: Was the `fs` argument needed? I got the impression from the docs that it would auto-infer the file system based on the URL's protocol.

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+        self.width = width
+  
+
+    def __getitem__(self, idx):
+        img = self.image_dataset[idx]
+
+
+        image = np.asarray(img['image'])
+        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB).astype(np.float32)
+        image_resized = cv2.resize(image, (self.width, self.height))
```
Comment (treubig26) [inline examples/src/charmory_examples/object_detection/xview/train_xview_model.py:159]: Could the resize have been done using albumentations?

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        collate_fn=collate_fn
+    )
+    return valid_loader
+import torchvision
+from torchvision.models.detection.faster_rcnn import FastRCNNPredictor,FasterRCNN_MobileNet_V3_Large_320_FPN_Weights
+
+model = torchvision.models.detection.fasterrcnn_mobilenet_v3_large_320_fpn(weights=FasterRCNN_MobileNet_V3_Large_320_FPN_Weights.DEFAULT)
+
+def create_model(num_classes):
+    num_classes = num_classes   # 63 classes
```
Comment (treubig26) [inline examples/src/charmory_examples/object_detection/xview/train_xview_model.py:232]: This line is not necessary or doing anything, right?

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+print(f"Number of validation samples: {len(test_dataset)}\n")
+# initialize the model and move to the computation device
+model = create_model(6)
+model = model.to(DEVICE)
+
+
+from tqdm.auto import tqdm
+
+# function for running training iterations
+def train(train_data_loader, model):
```
Comment (treubig26) [inline examples/src/charmory_examples/object_detection/xview/train_xview_model.py:253]: Is there a particular reason this is done manually instead of using a Lightning Trainer?

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    # start timer and carry out training and validation
+    start = time.time()
+    train_loss = train(train_loader, model)
+    val_loss = validate(valid_loader, model)
+    print(f"Epoch #{epoch+1} train loss: {train_loss_hist.value:.3f}")   
+    print(f"Epoch #{epoch+1} validation loss: {val_loss_hist.value:.3f}")   
+    end = time.time()
+    print(f"Took {((end - start) / 60):.3f} minutes for epoch {epoch}")
+
+    time.sleep(5)
```
Comment (treubig26) [inline examples/src/charmory_examples/object_detection/xview/train_xview_model.py:341]: Why is there a sleep here?

File: examples/src/charmory_examples/object_detection/xview/test_xview_model.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        ToTensorV2(p=1.0),
+    ], bbox_params={
+        'format': 'pascal_voc', 
+        'label_fields': ['labels']
+    })
+
+
+
+
+class xview(torch.utils.data.Dataset):
```
Comment (treubig26) [inline examples/src/charmory_examples/object_detection/xview/test_xview_model.py:70]: Is this class the same as the one in the other file?

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py (hunk 1 above)
Comment (Honaker) [reply examples/src/charmory_examples/object_detection/xview/train_xview_model.py:18]: I took this out, thanks. It was from when I was using the original xview dataset and there were 63 classes and the labels were not in order and skipped some numbers for some reason.

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py (hunk 2 above)
Comment (Honaker) [reply examples/src/charmory_examples/object_detection/xview/train_xview_model.py:85]: Good point, I need to take this out of my other files as well. I was just following the examples online but I guess this is not needed anymore.

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py (hunk 5 above)
Comment (Honaker) [reply examples/src/charmory_examples/object_detection/xview/train_xview_model.py:253]: I did try to do this, but the lightning file was a pretty similar structure as this file. Meaning I had to create another class that was a child class of the LightningModule that was the only main difference. I don't think I ever got it completely working b/c there were some issues that I couldn't figure out. I was able to get this one working pretty easily. Do you have a clean example that you have from one of your experiments?

Comment (Honaker) [reply examples/src/charmory_examples/object_detection/xview/train_xview_model.py:253]: I did try to do this, but the lightning file was a pretty similar structure as this file. Meaning I had to create another class that was a child class of the LightningModule that was the only main difference. I don't think I ever got it completely working b/c there were some issues that I couldn't figure out. I was able to get this one working pretty easily. Do you have a clean example that you have from one of your experiments? I could try to modify that file to get lighting training working

File: examples/src/charmory_examples/object_detection/xview/test_xview_model.py (hunk 7 above)
Comment (Honaker) [reply examples/src/charmory_examples/object_detection/xview/test_xview_model.py:70]: Yes thanks!

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+        image_height = image.shape[0]
+
+        boxes = img['objects']['bbox']
+        box_final = []
+        for box in boxes:
+            xmin_final = (box[0]/image_width)*self.width
+            xmax_final = (box[2]/image_width)*self.width
+            ymin_final = (box[1]/image_height)*self.height
+            yamx_final = (box[3]/image_height)*self.height
+            box_final.append([xmin_final, ymin_final, xmax_final, yamx_final])
```
[the remark this answers is not in the document]
Comment (Honaker) [reply examples/src/charmory_examples/object_detection/xview/train_xview_model.py:171]: I think that will just convert the bbox to a new format. What I am doing is just using changing the bbox locations since I resized the image

[the remark this answers is not in the document]
Comment (treubig26) [reply examples/src/charmory_examples/object_detection/xview/train_xview_model.py:171]: Ah. albumentations might be a cleaner way to handle resizing of the bounding boxes as well. I see we are using albumentations for other transforms, why is resizing not also handled there?

[the remark this answers is not in the document]
Comment (Honaker) [reply examples/src/charmory_examples/object_detection/xview/train_xview_model.py:171]: Thanks, sorry should have deleted that comment, realized that yesterday that I can just transform the dataset with albumentations and that should handle the resizing and I won't need to do much inside that method of the class. Currently trying to get that to work

File: examples/src/charmory_examples/object_detection/xview/train_xview_model.py (hunk 3 above)
Comment (Honaker) [reply examples/src/charmory_examples/object_detection/xview/train_xview_model.py]: Yes that should be fixed now

**Review by treubig26 — APPROVED** [review]
